backend/main/chapters/c11_tic_tac_toe_project.py

```diff
@@ -0,0 +1,606 @@
+- finally a `main` function that puts it all together and runs the game interactively.
+- Later on we will add further improvements.
+
+Let's get started!
+
+As in the last chapter, we will represent the tic-tac-toe board as a nested list of strings.
+For a typical game this will be a 3x3 list, i.e. 3 lists each containing 3 strings, with players represented by `'X'` or `'O'`.
+Empty squares will be represented by a space, i.e. `' '`. For example:
+
+    board = [
+        ['X', 'O', 'X'],
+        ['O', ' ', ' '],
+        [' ', 'X', 'O']
+    ]
+
+However to make things more interesting your code will need to work for square boards of any size
+(4x4, 5x5, etc) where players can be represented by any two strings, e.g.
+
+    board = [
+        ['A', 'B', 'A', 'A'],
+        ['B', ' ', ' ', 'A'],
+        [' ', 'A', 'B', 'B'],
+        [' ', 'A', 'B', ' ']
+    ]
+
+We want a function `row_winner`, which takes an argument `board` like above and
+returns `True` if `board` contains a winning row of either piece, `False` otherwise.
+
+One naive way to achieve this for 3x3 boards is:
+
+    def row_winner(board):
+        return ['O', 'O', 'O'] in board or ['X', 'X', 'X'] in board
+
+There are some issues with this solution:
+it won't work for pieces other than `'X'` and `'O'`, and it won't work for boards of other sizes.
+Also this approach won't work for `column_winner` or `diagonal_winner` later.
+
+I'd like you to find a better solution. Write a `row_winner` that works for any two
+different characters as pieces (except `' '` of course), and for boards of arbitrary size.
+A winning row would have the same character in all its entries (except `' '`).
+Here are some tests it should pass (feel free to add more):
+
+    __copyable__
+    assert_equal(
+        row_winner(
+            [
+                ['A', 'A', 'B', 'A'],
+                [' ', ' ', ' ', ' '],
+                ['A', ' ', ' ', 'A'],
+                ['B', ' ', 'B', 'A']
+            ]
+        ),
+        False
+    )
+    assert_equal(
+        row_winner(
+            [
+                ['X', ' ', 'X'],
+                ['O', 'X', 'X'],
+                ['O', 'O', 'O']
+            ]
+        ),
+        True
+    )
+
+        """
+
+        hints = """
+You need to check every row, so you'll need a loop.
+How can you check if all entries in a row are equal to each other?
+If you don't know the size of a row, you'll have to loop all the way through it.
+So you'll end up using nested loops!
+For each row, define a boolean. Then loop through that row, updating the boolean accordingly.
+You can use the first entry `row[0]` in a row to compare all the row entries to it.
+Think carefully about what the initial value of the boolean should be, and under what conditions you should change its value.
+After looping through a row, if you determined that all its entries are equal, then return `True` (ending the outer loop early).
```

Comment:
Need to cover this concept in the functions chapter

Reply:
OK. Will get to it. Would `Returning Values from Functions` be the best place for it? Would one `VerbatimStep` suffice? Or should I make a new page with many steps?

Reply:
On second thought, I should add at least 2 or 3 steps. I should show one `if-else` block example, maybe mention that `else` is unnecessary; and I should show one `for` loop example, with Snoop, so they can see the loop doesn't go all the way.